### ISSUE TITLE
Preflight request with OPTIONS method for token introspection endpoint not working.

### ISSUE THREAD

**Issue by syadhunandan** [keycloak/keycloak#29039]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

authentication

### Describe the bug

When using the http request for token introspection endpoint with OPTIONS method, it returns “**405 method not allowed**” in keycloak version 23 and 24.

This error happened only after the upgrade.

while checking with older versions till 22.0.5, http request for token introspection endpoint with OPTIONS method returns “200”

There was no mention about it in the release notes.

### Version

24.0.3

### Regression

- [X] The issue is a regression

### Expected behavior

For preflight http request with OPTIONS method for token introspection end point, it should return the response code as **200 OK**

### Actual behavior

For preflight http request with OPTIONS method for token introspection end point, it returns **405 method not allowed**

### How to Reproduce?

With keycloak newer versions 23 and later, 
_curl -X OPTIONS https://localhost:8080/auth/realms/{realm.name}/protocol/openid-connect/token/introspect_

**Comment by rmartinc** [comment]
Hi @syadhunandan!

You are right and this is because the change to resteasy reactive in version 23. Previously [this method](https://github.com/keycloak/keycloak/blob/24.0.3/services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java#L142-L145) seems to manage the options call. But now it seems to have another method in the [TokenIntrospectionEndpoint](https://github.com/keycloak/keycloak/blob/24.0.3/services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenIntrospectionEndpoint.java) class. The test should be something similar to this [method](https://github.com/keycloak/keycloak/blob/24.0.3/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/TokenEndpointCorsTest.java#L46-L55).

This seems to be an easy one. If someone wants to fix it just let me kknow. If not I'll try to provide a fix in the next days.

**Comment by syadhunandan** [comment]
Hi @rmartinc !
Thank you for the response. Please let me know when you provide the fix. Thanks

**Comment by rmartinc** [comment]
Hi @syadhunandan!

I started to look at this and I have changed my mind. The [TokenIntrospectionEndpoint](https://github.com/keycloak/keycloak/blob/24.0.3/services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenIntrospectionEndpoint.java) does not manage any cors at all. That makes sense because this endpoint (by spec) needs authentication (secret or jwt) and it's just allowed for confidential (non public) clients. Therefore it is not thought for JavaScript browser side clients, in which the secret/JWT to authenticate would also be public.  In summary, this endpoint is not for browser clients and that's why it has no cors at all. So, following this reasoning, the previous behavior was incorrect, it responded for preflight (this was not done on purpose) but then there was no more cors in normal POST calls. So my vote now it's to continue to do it like it's now, no cors at all (IMHO better than doing cors only for the options call).

How are you calling this endpoint from a browser that needs the cors preflight? Note that the secret or JWT if you are using a browser is public and anyone can guess it...

**Comment by rmartinc** [comment]
Closing this because of inactivity. As commented in the previous paragraph this endpoint is thought to be called by confidential clients with authentication, not by browsers, so cors are not expected in this endpoint.

**Comment by carillonator** [comment]
@rmartinc is it possible the same situation applies to `/realms/<realm>/protocol/openid-connect/auth` ? I have also recently upgraded from keycloak 22 to 24.0.5 and have found OPTIONS requests to that endpoint now return 405.

In my case a web page is left open past session cookie expiration. Further interaction causes a fetch, which gets redirected to keycloak for login, during which the browser does CORS preflight.

Thanks

**Comment by gpsmit** [comment]
@rmartinc we are potentially seeing the same thing @carillonator is seeing on` /realms/<realm>/protocol/openid-connect/auth`
Our key cloak version is 23.0.5.  Our backend initiates the login flow when a protected resources is requested by sending a 302 with location header back to the browser, the browser then performs preflight checks which fail with a 405.

**Comment by rmartinc** [comment]
AFAIK the auth endpoint is not protected by CORS because it's expected to be used via browser, the browser is redirected to it, not called directly by JS. Check for example this issue in another project with the same answer: https://github.com/ory/hydra/issues/1211.

And please, don't write here when even the endpoint is different. If you think that auth is buggy or it can be improved, please file a new issue. Although I think that auth should not be CORS protected as it's not used directly by JS.

**Comment by x-c4mp3r** [comment]
I read the whole discussion here and on the 1211 hydra ticket, but still don't know how to fix my problem which related to CORS too. In keycloak V22.0.4 everything worked fine, but with V23.0.0 /auth endpoint CORS errors appeared and none of the newer version is solving my problem.
As I see in the hydra ticket it's not allowed to XHR request make the auth begin V23.0.0, but why that was allowed in older versions and what to do to solve the problem?
My project "flow" is the fallowing (angular frontend + wildfly backend+ keycloak):
```
- user opens webpage fe.: https://testingmyapp.org/common-web (this is an angular frontend hosted from a jakarta ee war)
- in the war the OIDC.json is setted up using wildfly elytron, so redirects happen to keycloak
- user types username and password hit login and redirected to https://testingmyapp.org/common-web
- frontend calling a backend war endpoint with angular httpclient get request fe.: https://testingmyapp.org/admin-web/rest/session-data/ping
- backend admin-web war project has OIDC.json setted up so redirect happen to https://keycloak.testingmyapp.org/realms/testrealm/protocol/openid-connect/auth?response_type=code&client_id=admin-web&redirect_uri=https%3A%2F%testingmyapp.org%2Fadmin-web%2Frest%2Fsession-data%2Fping&state=xxxxx&scope=openid
- redirect got CORS error: Reason: CORS preflight response did not succeed. Status code: 405.
```
Users have multiple roles each role is calling the appropriate web backend like:
admin role -> admin-web
customer role -> customer-web
sso worked with this way from keycloak 10 to keycloak 22.04.

OIDC.json content is this:
```
{
  "realm": "${keycloak.realm}",
  "auth-server-url": "${keycloak.auth.server.url}",
  "ssl-required": "none",
  "resource": "common-web",
  "credentials": {
    "secret": "xxxxx"
  }
},
{
  "realm": "${keycloak.realm}",
  "auth-server-url": "${keycloak.auth.server.url}",
  "ssl-required": "none",
  "resource": "admin-web",
  "credentials": {
    "secret": "xxxxx"
  }
},
{
  "realm": "${keycloak.realm}",
  "auth-server-url": "${keycloak.auth.server.url}",
  "ssl-required": "none",
  "resource": "customer-web",
  "credentials": {
    "secret": "xxxxx"
  }
}
```
war ping endpoint is this:
```
@GET
    @Path("/ping")
    public Response ping() {
        return Response.ok("").build();
    }
```
web.xml for admin is (same on customer just admin strings replaced to customer):
```
<?xml version="1.0" encoding="UTF-8"?>
<web-app xmlns="https://jakarta.ee/xml/ns/jakartaee"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="https://jakarta.ee/xml/ns/jakartaee https://jakarta.ee/xml/ns/jakartaee/web-app_5_0.xsd"
         version="5.0">
	<display-name>admin.web</display-name>

    <session-config>
        <session-timeout>240</session-timeout>
    </session-config>

    <distributable/>

    <welcome-file-list>
        <welcome-file>/index.html</welcome-file>
    </welcome-file-list>

    <servlet>
        <servlet-name>jakarta.ws.rs.core.Application</servlet-name>
        <load-on-startup>1</load-on-startup>
    </servlet>

    <servlet-mapping>
        <servlet-name>jakarta.ws.rs.core.Application</servlet-name>
        <url-pattern>/rest/*</url-pattern>
    </servlet-mapping>

    <login-config>
        <auth-method>OIDC</auth-method>
    </login-config>

    <security-role>
        <role-name>ADMIN</role-name>
    </security-role>

    <security-constraint>
        <web-resource-collection>
            <web-resource-name>Admin</web-resource-name>
            <url-pattern>/*</url-pattern>
        </web-resource-collection>
        <web-resource-collection>
            <web-resource-name>AdminSessionData</web-resource-name>
            <url-pattern>/rest/session-data/*</url-pattern>
            <http-method-omission>OPTIONS</http-method-omission>
        </web-resource-collection>
        <auth-constraint>
            <role-name>ADMIN</role-name>
        </auth-constraint>
    </security-constraint>
```
Thanks in advance

**Comment by MrdUkk** [comment]
yes. i have this problem too. the client is lua-openresty-openidc
weborigin set to *  OR +  (any variant lead to http 405 method not allowed on keycloak part)

token refresh issued on /realms/test/protocol/openid-connect/auth?response_type=code&client_id=lua-resty-openidc
ended with 405 error code.

**Comment by clouded-minds** [comment]
Any workarounds that anyone has been able to come up with?

**Comment by MrdUkk** [comment]
on the reverse proxy that is before keycloak in http request processing , 
add 'fake-cors' processing for this specific URL endpoint. only
